### ISSUE TITLE
Delete Enabled but not allowed to delete

### ISSUE THREAD

**Issue by jwhitcraft** [docker/distribution#989]
Hello

I have a registry started with the following config

```yaml
redis:
  ports:
  - 6379/tcp
  image: redis
  restart: always
registry:
  restart: always
  ports:
  - 443:5000/tcp
  environment:
    REGISTRY_LOG_LEVEL: info
    REGISTRY_HTTP_TLS_CERTIFICATE: /certs/_.xxxxx.net.crt
    REGISTRY_HTTP_TLS_KEY: /certs/_.xxxxx.net.key
    REGISTRY_STORAGE_CACHE_BLOBDESCRIPTOR: redis
    REGISTRY_REDIS_ADDR: redis:6379
    REGISTRY_STORAGE_DELETE_ENABLED: true
    REGISTRY_STORAGE: s3
    REGISTRY_STORAGE_S3_ACCESSKEY: my_access_key
    REGISTRY_STORAGE_S3_BUCKET: my_bukkit
    REGISTRY_STORAGE_S3_REGION: us-east-1
    REGISTRY_STORAGE_S3_SECRETKEY: my_secret_key
    REGISTRY_STORAGE_S3_V4AUTH: true
    REGISTRY_STORAGE_S3_CHUNKSIZE: 5242880
  image: registry:2.1.1
  links:
    redis: redis
  volumes:
  - /home/ubuntu/registry/certs:/certs
```

But when I issue a delete http call to `https://my-registry.xyz/v2/hello-world/manifests/sha256:2b471673e2552749332efa8cac83d4635dce8c276637d630e4c85ac99b4700be`, I get the 405 not supported error.  Am i missing something?

**Comment by RichardScothern** [comment]
I suspect the `REGISTRY_STORAGE: s3` line is overriding the line above as it resets the entire `REGISTRY_STORAGE` config struct.

Can you move `REGISTRY_STORAGE_DELETE_ENABLED: true` below it and retry.

**Comment by jwhitcraft** [comment]
@RichardScothern,

I just tried this config and still not allowing me to delete

```yaml
redis:
  ports:
  - 6379/tcp
  image: redis
  restart: always
registry:
  restart: always
  ports:
  - 443:5000/tcp
  environment:
    REGISTRY_LOG_LEVEL: info
    REGISTRY_HTTP_TLS_CERTIFICATE: /certs/_.xxxxx.net.crt
    REGISTRY_HTTP_TLS_KEY: /certs/_.xxxxx.net.key
    REGISTRY_REDIS_ADDR: redis:6379
    REGISTRY_STORAGE: s3
    REGISTRY_STORAGE_S3_ACCESSKEY: my_access_key
    REGISTRY_STORAGE_S3_BUCKET: my_bukkit
    REGISTRY_STORAGE_S3_REGION: us-east-1
    REGISTRY_STORAGE_S3_SECRETKEY: my_secret_key
    REGISTRY_STORAGE_S3_V4AUTH: true
    REGISTRY_STORAGE_S3_CHUNKSIZE: 5242880
    REGISTRY_STORAGE_DELETE_ENABLED: true
    REGISTRY_STORAGE_CACHE_BLOBDESCRIPTOR: redis
  image: registry:2.1.1
  links:
    redis: redis
  volumes:
  - /home/ubuntu/registry/certs:/certs
```

**Comment by RichardScothern** [comment]
I'm not sure what is going on.  
`REGISTRY_STORAGE_DELETE_ENABLED
Can you somehow echo the

**Comment by beyondblog** [comment]
@RichardScothern  Hello
I also encountered the same problem
![image](https://cloud.githubusercontent.com/assets/3734587/10090203/3a374afe-635f-11e5-8e8b-7a726ba51b7a.png)

```
docker inspect 79420d14d9db
```
![1](https://cloud.githubusercontent.com/assets/3734587/10090285/5bba9d56-6360-11e5-8c5e-6e31826fedaa.png)

![image](https://cloud.githubusercontent.com/assets/3734587/10090299/87a1589c-6360-11e5-9492-574397ca0ad8.png)

**Comment by RichardScothern** [comment]
@beyondblog Can you confirm the environment variable is being set within the container?

**Comment by BrianBland** [comment]
This looks like an issue with yaml serialization/deserialization of the value `true` from the compose file. Compose is reading in `REGISTRY_STORAGE_DELETE_ENABLED` as `True` (booleans are capitalized in python) and then seems to be printing out the string value instead of re-serializing this. Could you try changing your compose file to use `REGISTRY_STORAGE_DELETE_ENABLED: "true"` instead?

**Comment by beyondblog** [comment]
@BrianBland 

yes i can
this my compose file 

```
registry:                                                                                                                   
  image: registry:2                                                                                                         
  restart: always                                                                                                           
  ports:                                                                                                                    
    - 5000:5000                                                                                                             
  environment:                                                                                                              
    REGISTRY_HTTP_TLS_CERTIFICATE: /certs/domain.crt                                                                        
    REGISTRY_HTTP_TLS_KEY: /certs/domain.key                                                                                
    REGISTRY_AUTH: htpasswd                                                                                                 
    REGISTRY_AUTH_HTPASSWD_PATH: /auth/htpasswd                                                                             
    REGISTRY_AUTH_HTPASSWD_REALM: Registry Realm                                                                            
    REGISTRY_STORAGE_DELETE_ENABLED: true                                                                                   
                                                                                                                            
  privileged: true                                                                                                          

  volumes:                                                                                                                  
    - /boot/data:/var/lib/registry                                                                                          
    - /data/private-registry/certs:/certs                                                                                   
    - /data/private-registry/auth:/auth
```

**Comment by jwhitcraft** [comment]
@BrianBland,

I just tried this

```yaml
redis:
  ports:
  - 6379/tcp
  image: redis
  restart: always
registry:
  restart: always
  ports:
  - 443:5000/tcp
  environment:
    REGISTRY_LOG_LEVEL: info
    REGISTRY_HTTP_TLS_CERTIFICATE: /certs/_.xxxxx.net.crt
    REGISTRY_HTTP_TLS_KEY: /certs/_.xxxxx.net.key
    REGISTRY_REDIS_ADDR: redis:6379
    REGISTRY_STORAGE: s3
    REGISTRY_STORAGE_S3_ACCESSKEY: my_access_key
    REGISTRY_STORAGE_S3_BUCKET: my_bukkit
    REGISTRY_STORAGE_S3_REGION: us-east-1
    REGISTRY_STORAGE_S3_SECRETKEY: my_secret_key
    REGISTRY_STORAGE_S3_V4AUTH: true
    REGISTRY_STORAGE_S3_CHUNKSIZE: 5242880
    REGISTRY_STORAGE_DELETE_ENABLED: "true"
    REGISTRY_STORAGE_CACHE_BLOBDESCRIPTOR: redis
  image: registry:2.1.1
  links:
    redis: redis
  volumes:
  - /home/ubuntu/registry/certs:/certs
```

And it still won't let me delete.

Here is my screenshot from the `docker inspect` command showing that it's now set to `=true` instead of `=True` like it was before.
![screenshot 2015-10-01 08 28 35](https://cloud.githubusercontent.com/assets/26738/10220557/7d39f2aa-6816-11e5-8591-aa15092b95bb.png)

```
time="2015-10-01T12:25:02.505115659Z" level=error msg="response completed with error" err.code=UNSUPPORTED err.message="The operation is unsupported."
``` is in the logs

in the meantime, i just converted all my environment variables into a `config.yml` file and i'm passing it in as a volume and deletes are now working.

**Comment by BrianBland** [comment]
Sorry, it looks like this is most likely related to #888, which has been merged and is slated to be included in v2.2.

In fact, I was actually wrong about the yaml parsing of `True` as well, as it seems like it's very flexible with regards to parsing standard data types (see http://www.yaml.org/spec/1.2/spec.html#schema/core/ for more details).

tl;dr Feel free to unquote your boolean `true`s in your compose files.

**Comment by RichardScothern** [comment]
Just checked, this is fixed in master (by #888).  We have enough bug fixes for a patch release including this.

**Comment by grexe** [comment]
awesome, we are facing the same issue and really need to save some space - cannot wait for the patch!
